### ISSUE TITLE
Slow dependency resolution

### ISSUE THREAD

**Issue by pitrou** [conda/conda#2208]
I'm not sure if that's considerd very slow, but the `conda create` command below spends most of its CPU time (25 s.) in the "solving package specifications" step:
```
$ conda config --get channels
--add channels 'defaults'
--add channels 'antoine'
--add channels 'numba'
$ time conda create --dry-run -n toto numba
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ........
Solving package specifications: ...........

Package plan for installation in environment /home/antoine/miniconda3/envs/toto:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    llvmlite-0.9.0             |           py35_0         6.9 MB
    numpy-1.10.4               |           py35_1         5.9 MB
    setuptools-20.2.2          |           py35_0         457 KB
    numba-0.24.0               |     np110py35_99         1.9 MB
    pip-8.1.0                  |           py35_0         1.6 MB
    ------------------------------------------------------------
                                           Total:        16.7 MB

The following NEW packages will be INSTALLED:

    llvmlite:   0.9.0-py35_0       
    mkl:        11.3.1-0           
    numba:      0.24.0-np110py35_99
    numpy:      1.10.4-py35_1      
    openssl:    1.0.2g-0           
    pip:        8.1.0-py35_0       
    python:     3.5.1-0            
    readline:   6.2-2              
    setuptools: 20.2.2-py35_0      
    sqlite:     3.9.2-0            
    tk:         8.5.18-0           
    wheel:      0.29.0-py35_0      
    xz:         5.0.5-1            
    zlib:       1.2.8-0            

Dry run: exiting

real	0m31.301s
user	0m25.764s
sys	0m0.340s
```

**Comment by kalefranz** [comment]
I agree that 25 sec here seems "too long."  @mcg1969 with the inputs being so simple, guessing you'll want to dig in here a bit?

**Comment by mcg1969** [comment]
Eh, that's just going to happen sometimes---far less frequently than before, mind you, but still sometimes. Best way for me to understand what's happening is to look at the entire output of a `--debug` run.

**Comment by mcg1969** [comment]
Numba incidentally is going to be one of the worst offenders, because y'all have _so many builds_ :-)

**Comment by kalefranz** [comment]
Makes sense to me that it's expected behavior for now given the context.

**Comment by mcg1969** [comment]
This is what happens if I do the same command with the default environments:
```
real	0m1.432s
user	0m1.024s
sys	0m0.092s
```

**Comment by mcg1969** [comment]
Do paste the `--debug` output when you can, @pitrou ... I'll probably be able to see the issue clearly (as will you while you watch it churn for 25 seconds!)

**Comment by mcg1969** [comment]
Well, I'm stupid, I can roughly reproduce this myself. 
```
0203-mgrant:~ mgrant$ time conda --debug create --dry-run -n toto -c numba -c antoine numba
...
real	0m20.447s
user	0m16.468s
sys	0m0.459s
```
Looking at the debug output, there are *1160* numba packages the solver is considering. This is compared to only 68 packages if you don't include the `antoine` and `numba` channels. I'm not too concerned here.

**Comment by mcg1969** [comment]
Ah, I see an opportunity to improve performance here. Not a bug per se but definitely something I could do better. In short, the performance on the "build number optimization" stage is tied to the _largest_ number of builds across all versions---even if the newest version doesn't have that many.

**Comment by kalefranz** [comment]
Figured it would end up being just a small tweak :)

**Comment by mcg1969** [comment]
Ha ha! How's this?
```
real	0m5.432s
user	0m2.257s
sys	0m0.168s
```

**Comment by pitrou** [comment]
Sounds great :-) I suppose you're not interested in debug output anymore?

**Comment by mcg1969** [comment]
No thanks!

**Comment by kalefranz** [comment]
resolved with #2209
